quint/src/runtime/compile.ts

```diff
@@ -175,8 +177,17 @@ export function compileDef(
 ): CompilationContext {
   // Define a new module list with the new definition in the last module,
   // ensuring the original object is not modified
-  const modules: QuintModule[] = [...state.modules] // Those are flat, but introducing `def` might make them non-flat
-  const lastModule: FlatModule = state.modules[state.modules.length - 1] // This is not modules.pop() to ensure flatness
+  const originalModules = [...state.originalModules]
+  const originalLastModule = originalModules.pop()
+  if (!originalLastModule) {
+    throw new Error('No original modules in state')
+  }
```

Comment:
iiuc, this violates a contract for this method which we should maybe document on the method?

Also, given that this check and the analogous one on line 191 are checking requirements of this function, my personal preference would be to group these both at the top of function as part of its "preamble". But I'm not stubborn on this point if you disagree.